src/nci_util.c

```diff
@@ -912,6 +1018,169 @@ nci_discovery_ntf_copy(
     }
 }
 
+NciModeParam*
+nci_mode_param_copy(
+    const NciModeParam* param,
+    NCI_MODE mode)
+{
+    if (G_LIKELY(param)) {
+        NciModeParam* copy = NULL;
+        gsize size = calculate_mode_param_fit_size(param, mode);
+        gsize copied = 0;
+
+        switch (mode) {
+        case NCI_MODE_ACTIVE_POLL_A:        /* fallthrough */
+        case NCI_MODE_PASSIVE_POLL_A:       /* fallthrough */
+        case NCI_MODE_ACTIVE_POLL_F:        /* fallthrough */
+        case NCI_MODE_PASSIVE_POLL_F:       /* fallthrough */
+        case NCI_MODE_PASSIVE_POLL_B:       /* fallthrough */
+        case NCI_MODE_ACTIVE_LISTEN_F:      /* fallthrough */
+        case NCI_MODE_PASSIVE_LISTEN_F:     /* fallthrough */
+            copy = (NciModeParam*) g_malloc0(size);
+            copied = nci_mode_param_inplace_copy(param, copy, mode);
+            if(!copied) {
+                g_free(copy);
+                copy = NULL;
+            }
+            return copy;
+        case NCI_MODE_PASSIVE_POLL_15693:
+        case NCI_MODE_PASSIVE_LISTEN_15693:
+            break;
+        case NCI_MODE_PASSIVE_LISTEN_A:     /* fallthrough */
+        case NCI_MODE_PASSIVE_LISTEN_B:     /* fallthrough */
+        case NCI_MODE_ACTIVE_LISTEN_A:
+            /* NCI 1.0 defines no parameters for A/B Listen modes */
+            return NULL;
+        }
+        GDEBUG("Unhandled activation mode 0x%02x", mode);
+        return NULL;
+    } else {
+        return NULL;
+    }
+}
+
+NciActivationParam*
+nci_activation_param_copy(
+    const NciActivationParam* param,
+    NCI_RF_INTERFACE intf,
+    NCI_MODE mode)
+{
+    if (G_LIKELY(param)) {
+        NciActivationParam* copy = NULL;
+        gsize size = G_ALIGN8(sizeof(NciActivationParam));
+        const NciActivationParamIsoDepPollA* iso_dep_poll_a = NULL;
+        const NciActivationParamIsoDepPollB* iso_dep_poll_b = NULL;
+        const NciActivationParamNfcDepPoll* nfc_dep_poll = NULL;
+        const NciActivationParamNfcDepListen* nfc_dep_listen = NULL;
+        switch (intf) {
+        case NCI_RF_INTERFACE_ISO_DEP:
+            switch (mode) {
+            case NCI_MODE_PASSIVE_POLL_A:
+            case NCI_MODE_ACTIVE_POLL_A:
+                iso_dep_poll_a = &param->iso_dep_poll_a;
+                g_assert(iso_dep_poll_a->t1.bytes ? iso_dep_poll_a->t1.size :
```

Comment:
And no `g_assert`s

Reply:
And `iso_dep_poll_a->t1.bytes ? iso_dep_poll_a->t1.size : TRUE` doesn't seem to make much sense to me, if any.

Basically I read it like this: if `iso_dep_poll_a->t1.bytes` is not NULL then we claim that `iso_dep_poll_a->t1.size` is non-zero. That's a questionable statement - non-NULL pointer pointing to memory block of zero size does make certain amount of sense, or at least it's definitely not impossible. It's a NULL pointer to a non-empty memory block that doesn't make sense. And then it gets even worse. If `iso_dep_poll_a->t1.bytes` is NULL then we assert TRUE. And that sounds like total nonsense. Why would anyone assert just TRUE. TRUE is true, there's nothing to assert there! At that point my brain just short-circuits))

`iso_dep_poll_a->t1.bytes || !iso_dep_poll_a->t1.size` would make a lot more sense. It reads as "either pointer is non-NULL (in which case size can be whatever) or size is zero". Short and clear statement. C-style.